File: admin/fleetcommander/collectors.py
# -*- coding: utf-8 -*-
# vi:ts=4 sw=4 sts=4

# Copyright (C) 2014 Red Hat, Inc.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the licence, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this program; if not, see <http://www.gnu.org/licenses/>.
#
# Authors: Alberto Ruiz <aruiz@redhat.com>
#          Oliver Gutiérrez <ogutierrez@redhat.com>
#

import json
import logging


class DummyCollector:
    """
    Dummy collector class
    """

    def __init__(self, *args, **kwargs):
        pass

    def handle_change(self, change):
        """
        Handle a change and store it
        """
        pass

    def dump_changes(self):
        """
        Dump all changes in a sorted list
        """
        return []

    def remember_selected(self, selected_keys):
        """
        Mark given changes as selected
        """
        pass

    def get_settings(self):
        """
        Return a list of selected changes
        """
        return []

    def merge_settings(self, profile_settings):
        """
        Merge given settings pair
        """
        return []


class BaseCollector(object):
    """
    Base collector class
    """

    COLLECTOR_NAME = 'basecollector'

    def __init__(self, db):
        """
        Class initialization
        """
        self.db = db
        self.settings = self.db.sessionsettings

    def get_key_from_change(self, change):
        """
        Return change key identifier
        """
        if 'key' in change:
            return change['key']

    def get_value_from_change(self, change):
        """
        Return change human readable value
        """
        if 'value' in change:
            return change['value']
        return 'Undefined'

    def handle_change(self, change):
        """
        Handle a change and store it
        """
        key = self.get_key_from_change(change)
        if key is not None:
            self.settings.update_setting(
                self.COLLECTOR_NAME,
                key,
                json.dumps(change))
        else:
            logging.error('%s: Given change is invalid: %s' % (
                self.COLLECTOR_NAME,
                json.dumps(change)
            ))

    def dump_changes(self):
        """
        Dump all changes in a sorted list
        """
        data = []
        changes = self.settings.get_for_collector(self.COLLECTOR_NAME)
        for key, change in sorted(changes.items()):
            value = self.get_value_from_change(json.loads(change))
            data.append([key, value])
        return data

    def remember_selected(self, selected_keys):
        """
        Mark given changes as selected
        """
        changes = self.settings.get_for_collector(self.COLLECTOR_NAME)
        selection = []
        for key in selected_keys:
            if key in changes:
                selection.append(key)

        if len(selection) > 0:
            self.settings.select_settings(self.COLLECTOR_NAME, selection)

    def get_settings(self):
        """
        Return a list of selected changes
        """
        changes = self.settings.get_for_collector(
            self.COLLECTOR_NAME, only_selected=True)
        keys = sorted(changes.keys())
        return [json.loads(changes[key]) for key in keys]

    def merge_settings(self, profile_settings):
        """
        Return a list of selected changes
        """
        current = self.get_settings()
        index = {}
        for changeset in [profile_settings, current]:
            for change in changeset:
                index[change["key"]] = change
        return index.values()


class GSettingsCollector(BaseCollector):
    """
    GSettings collector class
    """
    COLLECTOR_NAME = 'org.gnome.gsettings'


class LibreOfficeCollector(BaseCollector):
    """
    LibreOffice collector class
    """
    COLLECTOR_NAME = 'org.libreoffice.registry'


class NetworkManagerCollector(BaseCollector):
    """
    Network manager collector class
    """
    COLLECTOR_NAME = 'org.freedesktop.NetworkManager'

    def get_key_from_change(self, change):
        """
        Return change key identifier for NM connection
        """
        try:
            return change[u'uuid']
        except:
            return

    def get_value_from_change(self, change):
        """
        Return change human readable value for NM connection
        """
        try:
            return '%s - %s' % (change[u'type'], change[u'id']),
        except:
            return 'Undefined'

File: admin/cockpit/fleet-commander-admin/js/collectors.js
/*
* Copyright (C) 2014 Red Hat, Inc.
*
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU Lesser General Public
* License as published by the Free Software Foundation; either
* version 2.1 of the licence, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
* Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public
* License along with this program; if not, see <http://www.gnu.org/licenses/>.
*
* Author: Oliver Gutiérrez <ogutierrez@redhat.com>
*/

/*global $ */
/*global DEBUG */

"use strict";

function BaseCollector(namespace) {
    this.namespace = namespace;
    this.changes = {};
    this.key_name = 'key';
}

BaseCollector.prototype = {

    get_key_from_change: function (change) {
        if (change[this.key_name] !== undefined) {
            return change[this.key_name];
        }
        return undefined;
    },

    get_value_from_change: function (change) {
        if (change.value !== undefined) {
            return change.value;
        }
        return undefined;
    },

    handle_change: function (change) {
        if (DEBUG > 0) {
            console.log(
                'FC: Collector',
                this.namespace,
                'handling change',
                change
            );
        }
        var key = this.get_key_from_change(change);
        if (key !== undefined) {
            this.changes[key] = change;
        } else {
            if (DEBUG > 0) {
                console.log(
                    'FC: Collector',
                    this.namespace,
                    'can not handle change',
                    change
                );
            }
        }
    },

    dump_changes: function () {
        var self = this, changelist = [];
        $.each(this.changes, function (k, v) {
            changelist.push([k, self.get_value_from_change(v)]);
        });
        return changelist;
    },

    get_changeset: function (selected_keys) {
        var self = this, changeset = [];
        /*jslint unparam: true */
        $.each(selected_keys, function (ignore, key) {
            if (self.changes[key] !== undefined) {
                changeset.push(self.changes[key]);
            }
        });
        /*jslint unparam: false */
        return changeset;
    }
};

// Network Manager specific collector
function NMCollector(namespace) {
    BaseCollector.apply(this, namespace);
    this.key_name = 'uuid';
}

NMCollector.prototype = Object.create(BaseCollector.prototype);

NMCollector.prototype.get_value_from_change = function (change) {
    if (change.type !== undefined && change.id !== undefined) {
        return change.type + ' - ' + change.id;
    }
    return undefined;
};

// Firefox bookmarks specific collector
function FirefoxBookmarksCollector(namespace) {
    BaseCollector.apply(this, namespace);
}

FirefoxBookmarksCollector.prototype = Object.create(BaseCollector.prototype);

FirefoxBookmarksCollector.prototype.get_value_from_change = function (change) {
    return change.value.URL + ' - ' + change.value.Title;
};
